src/librbd/AioRequest.cc

```diff
@@ -187,7 +310,48 @@ namespace librbd {
 				 << m_object_image_extents << dendl;
 
 	  m_state = LIBRBD_AIO_WRITE_COPYUP;
-	  read_from_parent(m_object_image_extents);
+
+          if (is_cor(m_ictx, m_snap_id)) {
+            m_ictx->copyup_list_lock.Lock();
+            it = m_ictx->copyup_list.find(m_object_no);
+            if (it == m_ictx->copyup_list.end()) {
+              // If it is not in the list, create a CopyupRequest and wait for it.
+              CopyupRequest *new_req = new CopyupRequest(m_ictx, m_oid,
+                                                         m_object_no, false);
+              // make sure to wait on this CopyupRequest
+              new_req->append_request(this);
+              pair<uint64_t, CopyupRequest*> new_slot =
+                pair<uint64_t, CopyupRequest*>(m_object_no, new_req);
+              m_ictx->copyup_list.insert(new_slot);
+              m_ictx->copyup_list_lock.Unlock();
+
+              m_entire_object = &(m_ictx->copyup_list[m_object_no]->get_copyup_data());
+              ldout(m_ictx->cct, 20) << __func__ << " creating new Copyup for AioWrite, obj-"
+                                     << m_object_no << dendl;
+              m_ictx->copyup_list[m_object_no]->read_from_parent(m_object_image_extents);
+              ldout(m_ictx->cct, 20) << __func__ << " issuing read_from_parent" << dendl;
+            } else {
+              Mutex::Locker l3(it->second->get_lock());
+              if (it->second->is_ready()) {
```

Comment:
Again -- very small time window since basically after the whole object is read, the CopyupRequest object should be deleted.